pkg/scheduler/framework/plugins/nodevolumelimits/csi.go

```diff
@@ -72,13 +76,68 @@ func (pl *CSILimits) Name() string {
 	return CSIName
 }
 
-// EventsToRegister returns the possible events that may make a Pod
+// EventsToRegister returns the possible events that may make a Pod.
 // failed by this plugin schedulable.
 func (pl *CSILimits) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
 		{Event: framework.ClusterEvent{Resource: framework.CSINode, ActionType: framework.Add}},
-		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}},
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}, QueueingHintFn: pl.isSchedulableAfterPodDeleted},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add}},
+	}
+}
+
+func (pl *CSILimits) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	deletedPod, _, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		return framework.Queue, fmt.Errorf("unexpected objects in isSchedulableAfterPodDeleted: %w", err)
+	}
+
+	if len(deletedPod.Spec.Volumes) == 0 {
+		return framework.QueueSkip, nil
+	}
+
+	if deletedPod.Spec.NodeName == "" {
+		return framework.QueueSkip, nil
+	}
+
+	csiNode, err := pl.csiNodeLister.Get(deletedPod.Spec.NodeName)
+	if apierrors.IsNotFound(err) {
+		// When csiNode for a deleted Pod is not found, it doesn't make the pod schedulable
+		// because csiNode should be recreated again to let the pod go to this node.
+		logger.V(5).Info("A PVC for the Pod is not found, this Pod won't be schedulable until it's recreated", "pod", klog.KObj(pod), "namespace", pod.Namespace)
+		return framework.QueueSkip, nil
+	} else if err != nil {
+		return framework.Queue, fmt.Errorf("could not get a CSINode object: %w", err)
 	}
+
+	deletedPodAttachedVolumes := make(map[string]string)
+	if err := pl.filterAttachableVolumes(logger, deletedPod, csiNode, false /* existing pod */, deletedPodAttachedVolumes); err != nil {
```

Comment:
We have to somehow ignore PVC-not-found error inside filterAttachableVolumes - PVC attached to deletedPod may already be deleted, and a deleted PVC can be just ignored rather than erroring that out.

Reply:
> We have to somehow ignore PVC-not-found error inside filterAttachableVolumes

Sorry, ignore this. 
If `newPod` is `false`, it already ignores not found errors inside.
https://github.com/kubernetes/kubernetes/blob/efa738f31c2abd7aac611d3ce160b1418cb61937/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L265C3-L278C4

Reply:
@sanposhiho 
Good point. But I think we don't need to catch such events.

Do you want to care for the type of `persistentVolumeClaim` volume, right?
https://kubernetes.io/docs/concepts/storage/volumes/#persistentvolumeclaim

It is not worth rescheduling until a PVC that a scheduling pod needs becomes available or added.
Thus, we don't need to take care of PVC NotFound cases because PVC doesn't exist either way.

---

Or are there cases where the PVCs are not actually deleted if they are not found?

Reply:
> It is not worth rescheduling until a PVC that a scheduling pod needs becomes available or added.
Thus, we don't need to take care of PVC NotFound cases because PVC doesn't exist either way.

To elaborate my example above:
- `deletedPod` has PVC-A. But, PVC-A is already deleted. PVC-A refers to PV-A.
- `pod` has PVC-B. PVC-B refers to PV-A.

Because PVC-A is already deleted, it's ignored in `filterAttachableVolumes` and it results in the entire `isSchedulableAfterPodDeleted` returns `QueueSkip`. But, we should return `Queue` in this case actually, right? Or maybe I misinterpret some implementation.

Reply:
It's very yayakoshi... Isn't there any case at all where two different PVCs get the same [`volumeUniqueName`](https://github.com/kubernetes/kubernetes/blob/a83939eaf881ba5a8abfae92456e762ff92ccc99/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L294)? even when it [falls back to getting driver info from storage class](https://github.com/kubernetes/kubernetes/blob/a83939eaf881ba5a8abfae92456e762ff92ccc99/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L348-L360)?

Reply:
Muzukashi.. I'm not 100% sure, but may be yes...?

Reply:
I'm sensing that errors that cannot find PVCs should be ignored to address multiple PVC cases. What do you think?

Reply:
Should be ignored → Basically Yes **but _only if_** there isn't any case at all where two different PVCs get the same [volumeUniqueName](https://github.com/kubernetes/kubernetes/blob/a83939eaf881ba5a8abfae92456e762ff92ccc99/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L294).

Reply:
@sanposhiho Sorry, but I couldn't get why we need to take care of that special case. May I ask you to guide me? (If you like, please DM me on Slack 🙏 )

> should be ignored → Basically Yes but only if there isn't any case at all where two different PVCs get the same [volumeUniqueName](https://github.com/kubernetes/kubernetes/blob/a83939eaf881ba5a8abfae92456e762ff92ccc99/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L294).

Reply:
Supposing:
1. `pod` has PVC-a.
2. `deletedPod` has PVC-b. 
3. PVC-a and PVC-b get the same `volumeUniqueName`.

In this case, the result from `isSchedulableAfterPodDeleted` could change, depending on whether PVC-b is already deleted or still alive.

Reply:
@sanposhiho 
Thanks! I got it. But when PVC-b has already been deleted, we should return QSkip because then PVC-a doesn't have a potential PVC to be able to reuse. After that, when a PVC has the same `volumeLimitKey` with `pod` added(probably creating margin thanks to deleting PVC-b), we can catch with ` {Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add}}`.

On the other hand, If PVC-b still exists when deletedPod is deleted, `pod` could use PVC-b. In that case, we should return `Queue` to attempt rescheduling.

Reply:
> PVC-a doesn't have a potential PVC to be able to reuse.

What do you mean by ^?

Reply:
In above example, regardless of whether PVC-b is still alive/deleted, deleting `deletedPod` (could) makes `pod` schedulable, right? or I may miss something

Reply:
> What do you mean by ^?

Let me rephrase 🙏 

PVC-b does not exist, i.e. it has been deleted, so there is no PVC available to `pod`, so it is not worth rescheduling

Reply:
Why does `pod` need PVC-b? It only uses PVC-a.

Reply:
> In above example, regardless of whether PVC-b is still alive/deleted, deleting `deletedPod` (could) makes `pod` schedulable, right? or I may miss something

My understanding is that the PVC must have been created in some way before the `pod` can be rescheduled?

Reply:
Yes. I meant, how is PVC-b required for `pod`? PVC-b has completely nothing to do with `pod`. 
PVC-b is just a PVC which is(was) used by `deletedPod`.

Reply:
Registering PVC/delete events is the only way to catch such a case.
https://github.com/kubernetes/kubernetes/pull/121508/commits/143114d9936e1fe4360fc2d269cfb1fa8c7124b1

Reply:
This explains that we indeed need to watch PVC/delete events, but is there any way to filter them?

Reply:
Maybe something as simple as being part of the same namespace? But ideally we can do more.

Reply:
@alculquicondor Yes, we can, but we should implement its filter in another PR. Should I create an issue for it?

Reply:
why in a different PR?

Reply:
@alculquicondor I'd like to keep this PR as small as possible.
https://github.com/kubernetes/kubernetes/issues/118893#issuecomment-1879899536

Reply:
To help me decide whether it makes sense to merge this PR standalone:
Is the new event matcher for PersistentVolumeClaim/Delete required only if you add the scheduling hint to Pod Delete?

Reply:
> Is the new event matcher for PersistentVolumeClaim/Delete required only if you add the scheduling hint to Pod Delete?

Yes, you are right. PVC/delete event is only helpful when QHint for Pod/Delete I added can't look up the PVC related to deleted pods.

Reply:
in that case, we should only add the event matcher when the scheduling hint feature gate is enabled.
Once that's done, I'm ok merging the PR.